README.md
serviceberry-jwt
=================

[![CircleCI](https://circleci.com/gh/bob-gray/serviceberry-jwt.svg?style=svg)](https://circleci.com/gh/bob-gray/serviceberry-jwt)
[![Test Coverage](https://api.codeclimate.com/v1/badges/354149574b260178e1ce/test_coverage)](https://codeclimate.com/github/bob-gray/serviceberry-jwt/test_coverage)
[![Maintainability](https://api.codeclimate.com/v1/badges/354149574b260178e1ce/maintainability)](https://codeclimate.com/github/bob-gray/serviceberry-jwt/maintainability)
[![npm version](https://badge.fury.io/js/serviceberry-jwt.svg)](https://badge.fury.io/js/serviceberry-jwt)

JSON Web Token plugin for [Serviceberry](https://serviceberry.js.org). For information
about JSON Web Tokens see [RFC 7519](https://tools.ietf.org/html/rfc7519).

Install
-------

```shell-script
npm install serviceberry-jwt
```

Usage
-----

Fails the request with a `401 Unauthorized` status if the token is not found, is not formatted correctly, or
if the key doesn't verify the signature.

This plugin exports an abstract class `Jwt` for extending by your authorization class that knows how to fetch your
token keys. To use this plugin extend `Jwt` and implement at least `getKey(id)`. The key is then used to verify
the request's token. You can extend the validation by overriding `validate(request)` or change the process of finding
a request's token by overriding `getToken(request)`.

```js
const Jwt = require("serviceberry-jwt");

class Auth extends Jwt {
	getKey (id) {
		return data.getKey(id); // can also return a promise or use async/await
	}

	async validate (request) { // you can override validate to go beyond just verifying the token signature
		await super.validate(...arguments); // will throw if token is not verified
		... // test scopes or other custom payload properties
	}
}

trunk.use(new Auth(verifyOptions, options));
```

verifyOptions *object*
----------------------

See [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken#user-content-jwtverifytoken-secretorpublickey-options-callback)
for verify options.

options *object*
----------------

  - **scheme** *string*

    The authentication scheme. Used to get the token from the request and to set the `WWW-Authenticate`
	response header scheme when responding with `401 Unauthorized`. Defaults to `Bearer`.

	  - *Bearer*

	    When the scheme is *Bearer* the plugin will look for the token in the request's
		`Authorization` header.

      - *Token*

	    When the scheme is *Token* the plugin will look for the token in the request's
		parameters by it's name (see *param* below). `request.getParam(options.param)`.

  - **accessToken** *boolean*

    When *scheme* is `Bearer` and `accessToken` is `true`, first look for the token in the `Authorization` header
    and if it isn't found look for `access_token` as described below. Defaults to `false`. This option has no
    purpose if *scheme* is `Token`. Per [RFC 6750](https://tools.ietf.org/html/rfc6750) sections 2.2 & 2.3.
    
      - Find bearer token in a form encoded body parameter named `access_token` for request methods with defined
        body semantics (*POST, PUT, and PATCH*).

      - Find bearer token in a query string parameter named `access_token` for request methods without defined
        body semantics (*not POST, PUT, or PATCH*).

  - **param** *string*

    When *scheme* is `Token`, *param* is the name of the request parameter where the token should be found.
    Defaults to `token`. This can be any type of request parameter - path, query string, or body. This option
    has no purpose if scheme is `Bearer`.

Jwt
---
Abstract class

### constructor([verifyOptions[, options]])

  - **verifyOptions** *object*

	Sets `this.verifyOptions`. Passed to `jwt.verify()`. See
	[jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken#user-content-jwtverifytoken-secretorpublickey-options-callback)

  - **options**

    Sets `this.options`. See [options](#options) above.

### getKey(id)

**You must extend this class and at least implement this method.**

Called by the `validate` method for fetching a signing key used to verify a token.

  - **id** *string*

    The id which identifies the key to be used to verify the token.

### use(request, response)

The [handler](https://serviceberry.js.org/docs/handlers) method. This is the method called by Serviceberry.
Sets `request.jwt`. This is an `async` function.

  - **request** *object*

    Serviceberry [`request`](https://serviceberry.js.org/docs/request) object.

  - **response** *object*

    Serviceberry [`response`](https://serviceberry.js.org/docs/response) object.

### validate(request)

Called by the `use` method to validate the token. This is an `async` function and should return
or resolve to a boolean value or throw an error.

  - **request** *object*

    Serviceberry [`request`](https://serviceberry.js.org/docs/request) object.

### getToken(request)

Called by the `use` method to find the request's token.

  - **request** *object*

    Serviceberry [`request`](https://serviceberry.js.org/docs/request) object.


CHANGELOG.md
0.3.0
-----
  - Updating [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken/wiki/Migration-Notes:-v8-to-v9) to version 9

0.2.1
-----
*No changes*

0.2.0
-----
  - Stopped setting `request.token` to `request.jwt.payload`
  - Deep-freezing `request.jwt` and setting its prototype to `null`
  - Added *accessToken* option
  - Throw error (calling `request.fail`) if decode is unsuccessful

spec/plugin.js
/* eslint-env jasmine */
/* eslint max-nested-callbacks: ["error", 3], no-return-assign: "warn" */

"use strict";

const Jwt = require("../plugin");

describe("serviceberry-jwt", () => {
	var handler,
		request,
		token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9." +
			"eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ." +
			"XbPfbIHMI6arZ3Y922BhjWgQzWXcXNrz0ogtVhfEd2o",
		errorHeaders = {
			"WWW-Authenticate": "Bearer"
		};

	beforeEach(() => {
		handler = new Jwt();
		handler.getKey = jasmine.createSpy("Jwt.getKey").and.returnValue("secret");
		request = createRequest("Bearer " + token);
	});

	it("should create a handler instance with a use() method", () => {
		expect(typeof handler.use).toBe("function");
	});

	it("should authenticate with a valid token", async () => {
		await handler.use(request);
	});

	it("should set request.jwt", async () => {
		await handler.use(request);

		expect(request.jwt).toEqual({
			string: request.getHeader("Authorization").replace("Bearer ", ""),
			header: {
				alg: "HS256",
				typ: "JWT"
			},
			payload: {
				sub: "1234567890",
				name: "John Doe",
				iat: 1516239022
			},
			signature: "XbPfbIHMI6arZ3Y922BhjWgQzWXcXNrz0ogtVhfEd2o"
		});
	});

	it("should deep freeze request.jwt", async () => {
		await handler.use(request);

		expect(() => request.jwt.header = "evil").toThrow();
		expect(() => request.jwt.payload.sub = "evil").toThrow();
		expect(() => request.jwt.signature = "evil").toThrow();
	});

	it("should not authenticate without credentials", async () => {
		request = createRequest();

		try {
			await handler.use(request);
		} catch {
			// request fail will throw an error
		}

		expect(request.fail).toHaveBeenCalledWith("Please provide token.", "Unauthorized", errorHeaders);
	});

	it("should not authenticate with the wrong key", async () => {
		handler.getKey.and.returnValue("wrong key");

		try {
			await handler.use(request);
		} catch {
			// request.fail will throw an error
		}

		expect(request.fail).toHaveBeenCalledWith("invalid signature", "Unauthorized", errorHeaders);
	});

	it("should not authenticate with a malformed token", async () => {
		request = createRequest("Bearer malformed");

		try {
			await handler.use(request);
		} catch {
			// request.fail will throw an error
		}

		expect(request.fail).toHaveBeenCalledWith("Token is malformed.", "Unauthorized", errorHeaders);
	});

	it("should get the token from the request's params if scheme is `Token`", () => {
		request.getParam.and.returnValue(token);

		handler = new Jwt({}, {
			scheme: "Token"
		});

		expect(handler.getToken(request)).toBe(token);
	});

	it("should get the token from the request's query string when accessToken is true", () => {
		request = createRequest();
		request.getQueryParam.and.returnValue(token);

		handler = new Jwt({}, {
			scheme: "Bearer",
			accessToken: true
		});

		expect(handler.getToken(request)).toBe(token);
		expect(request.getQueryParam).toHaveBeenCalledWith("access_token");
	});

	it("should get the token from the request's query string when accessToken is true", () => {
		request = createRequest();
		request.getMethod.and.returnValue("POST");
		request.getContentType.and.returnValue("application/x-www-form-urlencoded");
		request.getBodyParam.and.returnValue(token);

		handler = new Jwt({}, {
			scheme: "Bearer",
			accessToken: true
		});

		expect(handler.getToken(request)).toBe(token);
		expect(request.getBodyParam).toHaveBeenCalledWith("access_token");
	});

	it("should get the token from the authorization header first when accessToken is true", () => {
		request = createRequest(token);

		handler = new Jwt({}, {
			scheme: "Bearer",
			accessToken: true
		});

		expect(handler.getToken(request)).toBe(token);
		expect(request.getBodyParam).not.toHaveBeenCalled();
		expect(request.getQueryParam).not.toHaveBeenCalled();
	});

	it("should not find a token when the scheme is unknown", () => {
		handler = new Jwt({}, {
			scheme: "Foo Baz"
		});

		expect(handler.getToken(request)).toBeUndefined();
	});

	it("should fail if abstract method getKey is not overridden with an implementation", async () => {
		handler = new Jwt();

		try {
			await handler.use(request);
			fail();
		} catch (error) {
			expect().nothing();
		}
	});
});

function createRequest (authorization) {
	var request = jasmine.createSpyObj("request", [
		"getHeader",
		"getMethod",
		"getContentType",
		"getParam",
		"getBodyParam",
		"getQueryParam",
		"fail"
	]);

	request.getHeader.and.returnValue(authorization);
	request.fail.and.throwError("request fail");

	return request;
}


plugin.js
"use strict";

const jwt = require("jsonwebtoken"),
	{promisify} = require("util"),
	verify = promisify(jwt.verify),
	bearer = "Bearer ",
	methodsWithBody = ["POST", "PUT", "PATCH"];

class Jwt {
	constructor ({...verifyOptions} = {}, {...options} = {scheme: "Bearer"}) {
		Object.assign(this, {verifyOptions, options});

		if (this.options.scheme === "Token" && !this.options.param) {
			this.options.param = "token";
		}
	}

	async use (request) {
		var token = this.getToken(request);

		if (!token) {
			this.unauthorized(request, "Please provide token.");
		}

		request.jwt = Object.assign(Object.create(null), jwt.decode(token, {
			complete: true
		}));

		deepFreeze(Object.assign(request.jwt, {
			string: token
		}));

		if (!request.jwt.payload) {
			this.unauthorized(request, "Token is malformed.");
		}

		return this.validate(request);
	}

	async validate (request) {
		const key = await this.getKey(request.jwt.header.kid);

		try {
			await verify(request.jwt.string, key, this.verifyOptions);
		} catch (error) {
			this.unauthorized(request, error.message);
		}
	}

	getToken (request) {
		var {scheme, accessToken, param} = this.options,
			token;

		if (scheme === "Bearer" && accessToken) {
			token = this.getBearerToken(request);
		} else if (scheme === "Bearer") {
			token = this.getAuthHeaderToken(request);
		} else if (scheme === "Token") {
			token = request.getParam(param);
		}

		return token;
	}

	getBearerToken (request) {
		var token = this.getAuthHeaderToken(request);

		// Per [RFC 6750](https://tools.ietf.org/html/rfc6750) sections 2.2 & 2.3
		// bearer tokens can be passed as access_token in the body if body is form encoded and method has body
		if (!token && methodsWithBody.includes(request.getMethod()) &&
			request.getContentType() === "application/x-www-form-urlencoded") {
			token = request.getBodyParam("access_token");

		// or as access_token in the query string
		} else if (!token) {
			token = request.getQueryParam("access_token");
		}

		return  token;
	}

	getAuthHeaderToken (request) {
		var token = request.getHeader("Authorization") || "";

		if (token.startsWith(bearer)) {
			token = token.slice(bearer.length);
		}

		return token;
	}

	async getKey () {
		throw new Error(
			"serviceberry-jwt plugin exports an abstract class (Jwt). " +
			"Consumers of the plugin must extend this class and at least implement the getKey(id) method."
		);
	}

	unauthorized (request, message) {
		request.fail(message, "Unauthorized", {
			"WWW-Authenticate": this.options.scheme
		});
	}
}

function deepFreeze (value) {
	if (value === Object(value)) {
		Object.getOwnPropertyNames(Object.freeze(value))
			.forEach(name => deepFreeze(value[name]));
	}
}

module.exports = Jwt;
